firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigAutoFetch.java

```diff
@@ -0,0 +1,175 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.remoteconfig.internal;
+
+import static com.google.firebase.remoteconfig.FirebaseRemoteConfig.TAG;
+
+import android.util.Log;
+import com.google.android.gms.tasks.Task;
+import com.google.android.gms.tasks.Tasks;
+import com.google.firebase.remoteconfig.ConfigUpdateListener;
+import com.google.firebase.remoteconfig.FirebaseRemoteConfigRealtimeUpdateFetchException;
+import com.google.firebase.remoteconfig.FirebaseRemoteConfigRealtimeUpdateStreamException;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.net.HttpURLConnection;
+import java.util.Random;
+import java.util.Set;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+public class ConfigAutoFetch {
+
+  private static final int FETCH_RETRY = 3;
+
+  private final Set<ConfigUpdateListener> eventListener;
+
+  private final HttpURLConnection httpURLConnection;
+  private final ConfigFetchHandler configFetchHandler;
+  private final ConfigUpdateListener retryCallback;
+  private final ScheduledExecutorService scheduledExecutorService;
+  private final Random random;
+  private final Executor executor;
+
+  public ConfigAutoFetch(
+      HttpURLConnection httpURLConnection,
+      ConfigFetchHandler configFetchHandler,
+      Set<ConfigUpdateListener> eventListener,
+      ConfigUpdateListener retryCallback,
+      Executor executor,
+      ScheduledExecutorService scheduledExecutorService) {
+    this.httpURLConnection = httpURLConnection;
+    this.configFetchHandler = configFetchHandler;
+    this.eventListener = eventListener;
+    this.retryCallback = retryCallback;
+    this.scheduledExecutorService = scheduledExecutorService;
+    this.random = new Random();
+    this.executor = executor;
+  }
+
+  public void beginAutoFetch() {
+    executor.execute(
+        new Runnable() {
+          @Override
+          public void run() {
+            listenForNotifications();
+          }
+        });
+  }
+
+  // Check connection and establish InputStream
+  private void listenForNotifications() {
+    if (httpURLConnection != null) {
+      try {
+        int responseCode = httpURLConnection.getResponseCode();
+        if (responseCode == 200) {
+          InputStream inputStream = httpURLConnection.getInputStream();
+          handleNotifications(inputStream);
+          inputStream.close();
+        } else {
+          for (ConfigUpdateListener listener : eventListener) {
```

Comment:
Nit: Maybe it would make sense to refactor this block to a method like `broadcastErrorToListeners` since it's used in several places.